File: luogu/P2704.py
# -*- coding: utf-8 -*-
# ======================================
# @File    : P2704.py
# @Time    : 2020/2/9 14:21
# @Author  : Rivarrl
# ======================================
import sys
# P2704 [NOI2001]炮兵阵地
# 状态压缩dp
# 状态与上两行的相关
# 将摆放士兵状态设为1，否则为0
# 两行相与不是0时存在互相攻击的炮兵
# dp[i][j][k]表示第i行状态为j，第i-1行状态为k时的最大炮兵数

def solve(arr, n, m):
    # sta保存状态，count保存该状态号对应状态的炮兵数(1的个数)
    sta, count = [0] * 70, [0] * 70
    length = 0
    # 1024^2*100枚举范围过大，需要预处理，枚举行内符合条件的状态保存到sta中
    for i in range(1 << m):
        if (i & (i << 1)) or (i & (i << 2)): continue
        sta[length] = k = i
        while k > 0:
            count[length] += k & 1
            k >>= 1
        length += 1
    dp = [[[0] * length for _ in range(length)] for _ in range(n)]
    # 初始化dp[0][i][0]
    for i in range(length):
        # 不能把炮兵放在山上
        if sta[i] & arr[0]: continue
        dp[0][i][0] = count[i]
    # 初始化dp[1][i][j]
    for i in range(length):
        if sta[i] & arr[1]: continue
        for j in range(length):
            if sta[j] & arr[0]: continue
            if sta[i] & sta[j]: continue
            dp[1][i][j] = max(dp[1][i][j], dp[0][j][0] + count[i])
    for row in range(2, n):
        # 枚举当前第row行状态
        for i in range(length):
            if sta[i] & arr[row]: continue
            # 枚举row-1行的状态
            for j in range(length):
                # 不能把炮兵放在山上
                if sta[j] & arr[row-1]: continue
                # 不能和row行的打起来
                if sta[i] & sta[j]: continue
                # 枚举row-2行的状态
                for k in range(length):
                    # 不能把炮兵放在山上
                    if sta[k] & arr[row-2]: continue
                    # 不能打架
                    if (sta[i] & sta[k]) or (sta[j] & sta[k]): continue
                    # 将row-1行合理的布阵结果dp[row-1][j][k]加上第row行的布阵数
                    dp[row][i][j] = max(dp[row][i][j], dp[row-1][j][k] + count[i])
    res = 0
    for i in range(length):
        for j in range(length):
            res = max(res, dp[n-1][i][j])
    return res

if __name__ == '__main__':
    line1 = sys.stdin.readline().strip()
    n, m = map(int, line1.split(' '))
    arr = [0] * n
    for i in range(n):
        line = sys.stdin.readline().strip()
        for j in range(m):
            # 山地用1，平地用0
            # 有山的位置不可以摆放炮兵，等价于摆放过炮兵的状态1
            if line[j] == 'H':
                arr[i] |= 1 << j
    res = solve(arr, n, m)
    print(res)

File: luogu/P2704.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#define MAXSTATE 70 // 单行最多100个位置时的行内满足条件的状态最大估计值
#define MAXROW 101
#define MAXCOL 11
#define max(a,b) a>b?a:b
#define mset(a) memset(a,0,sizeof(a))
using namespace std;

int n, m;
int arr[MAXROW]; // 记录地形，平地：0，山坡：1
int sta[MAXSTATE]; // 记录行内满足条件的状态, 摆放炮兵：1，不摆放：0
int count[MAXSTATE]; //记录每种状态对应的1的个数
int dp[MAXROW][MAXSTATE][MAXSTATE]; // dp[i][j][k]表示第i行为状态j且第i-1行为k时的最优解
char ch;
int main() {
    mset(arr), mset(sta), mset(count), mset(dp);
    cin >> n >> m;
    for (int i=0; i<n; i++) {
        for (int j=0; j<m; j++) {
            cin >> ch;
            // 山坡不能摆放炮兵，相当于已经摆放的状态：1
            if (ch == 'H') { 
                arr[i] |= (1 << j);
            }
        }
    }
    int c = 0;
    // 预处理，把所有一行内符合条件的布阵状态记录到sta，对于的1的个数记录到count
    for (int i=0; i<(1<<m); i++) {
        if ((i & (i << 1)) || (i & (i << 2))) continue;
        int k = i;
        while (k > 0) {
            count[c] += (k & 1);
            k >>= 1;
        }
        sta[c++] = i;
    }
    // 初始化第1行
    for (int i=0; i<c; i++) {
        if (sta[i] & arr[0]) continue;
        dp[0][i][0] = count[i];
    }
    // 初始化第2行
    for (int i=0; i<c; i++) {
        if (sta[i] & arr[1]) continue;
        for (int j=0; j<c; j++) {
            if (sta[j] & arr[0]) continue;
            if (sta[i] & sta[j]) continue;
            dp[1][i][j] = max(dp[1][i][j], dp[0][j][0] + count[i]);
        }
    }
    // 从第3行开始dp
    for (int i=2; i<n; i++) {
        for (int j=0; j<c; j++) {
            // 炮兵不能上山
            if (sta[j] & arr[i]) continue;
            for (int k=0; k<c; k++) {
                if (sta[k] & arr[i-1]) continue;
                // 两行的炮兵互相伤害
                if (sta[j] & sta[k]) continue;
                for (int p=0; p<c; p++) {
                    if (sta[p] & arr[i-2]) continue;
                    if ((sta[j] & sta[p]) || (sta[k] & sta[p])) continue;
                    dp[i][j][k] = max(dp[i][j][k], dp[i-1][k][p] + count[j]);
                }
            }
        }
    }
    int res = 0;
    for (int i=0; i<c; i++) {
        for (int j=0; j<c; j++) {
            res = max(res, dp[n-1][i][j]);
        }
    }
    cout << res;
    return 0;
}